src/libraries/Microsoft.Extensions.Logging/gen/TemplateHelper.cs

```diff
@@ -0,0 +1,346 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Text
+{
+    internal static class TemplateHelper
+    {
+        /// <summary>
+        /// Parses the message template and throws exception for malformed messages.
+        /// </summary>
+        internal static string Parse(this string msg, List<string>? templates)
+        {
+            ReadOnlySpan<char> format = msg.AsSpan();
+            var builder = new StringBuilder(msg.Length);
+            var pos = 0;
+            var len = format.Length;
+            var ch = '\0';
+            var segments = new List<Segment>();
+            var numArgs = 0;
+
+            while (true)
+            {
+                var segStart = builder.Length;
+                while (pos < len)
+                {
+                    ch = format[pos];
+
+                    pos++;
+                    if (ch == '}')
+                    {
+                        if (pos < len && format[pos] == '}')
+                        {
+                            // double }, treat as escape sequence
+                            pos++;
+                        }
+                        else
+                        {
+                            // dangling }, fail
+                            throw new ArgumentException($"Dangling }} in format string at position {pos}", nameof(format));
```

Comment:
I assume these strings will be shown to the developer. If so, we should put them in a resource so they can be localized.